GDJS/GDJS/Extensions/Builtin/CommonInstructionsExtension.cpp

```diff
@@ -101,6 +103,72 @@ CommonInstructionsExtension::CommonInstructionsExtension() {
         return outputCode;
       });
 
+  GetAllEvents()["BuiltinCommonInstructions::Async"].SetCodeGenerator(
+      [](gd::BaseEvent& event_,
+         gd::EventsCodeGenerator& codeGenerator,
+         gd::EventsCodeGenerationContext& parentContext) {
+        gd::AsyncEvent& event = dynamic_cast<gd::AsyncEvent&>(event_);
+
+        const gd::String callbackFunctionName =
+            "asyncCallback" +
+            gd::String::From(codeGenerator.GenerateSingleUsageUniqueIdFor(
+                &event.GetInstruction()));
+
+        // Generate callback code
+        gd::EventsCodeGenerationContext callbackContext;
+        callbackContext.AsyncInheritsFrom(parentContext);
+
+        // Generate actions
+        gd::String actionsCode = codeGenerator.GenerateActionsListCode(
+            event.GetActions(), callbackContext);
+
+        // Generate subevents
+        if (event.HasSubEvents())  // Sub events
+        {
+          actionsCode += "\n{ //Subevents\n";
+          actionsCode += codeGenerator.GenerateEventsListCode(
+              event.GetSubEvents(), callbackContext);
+          actionsCode += "} //End of subevents\n";
+        }
+
+        // Compose the callback function and add outside main
+        const gd::String actionsDeclarationsCode =
+            codeGenerator.GenerateObjectsDeclarationCode(callbackContext);
+        const gd::String callbackCode = "function " + callbackFunctionName +
+                                        "(runtimeScene, asyncObjectsList) {\n" +
+                                        actionsDeclarationsCode + actionsCode +
+                                        "}\n";
+        codeGenerator.AddCustomCodeOutsideMain(callbackCode);
+
+        // Generate code to backup the objects lists
+        gd::String objectsListsCode =
+            parentContext.IsAsync()
+                ? "asyncObjectsList = "
+                  "gdjs.LongLivedObjectsList.from(asyncObjectsList);\n"
+                : "const asyncObjectsList = new gdjs.LongLivedObjectsList();\n";
+        // The objects to be declared of the callback context tell all objects
+        // used in the callbacks so that they can be backed up now.
+        for (const gd::String& objectToBackup :
+             callbackContext.GetAllDeclaredObjectsAcrossChildren()) {
+          if (parentContext.ObjectAlreadyDeclared(objectToBackup))
+            objectsListsCode +=
+                "for (const obj of " +
+                codeGenerator.GetObjectListName(objectToBackup, parentContext) +
+                ") asyncObjectsList.addObject(obj);\n";
+        }
+
+        // Generate the action with a .then to the generated callback
+        const gd::String actionCode = codeGenerator.GenerateActionCode(
+            event.GetInstruction(),
+            parentContext,
+            "() => (" + callbackFunctionName + "(" +
+                dynamic_cast<gdjs::EventsCodeGenerator&>(codeGenerator)
```

Comment:
That kind of sucks, but should be fine since this is a code generator for the GDJS platform exclusively.